### PR TITLE
Transit: Support signing a CSR template and setting a certificate chain

### PR DESCRIPTION
**Pull request by Gabrielopesantos**
This PR adds two new endpoints to the Transit engine.
- `keys/:name/csr` signs a CSR using the provided key, ensuring the key material stays within Transit. If no CSR is provided, it signs an empty CSR. Otherwise, it signs the provided CSR, replacing its key material with the key material. The key in transit must be a signing key and not be derived.

- `keys/:name/set-certificate` sets the certificate chain for the provided key, ensuring the key material stays within Transit and certificates are managed in one place. It also allows chain updates and rotation, as it will overwrite any existing certificate chain.

Resolves #

## Target Release

1.14.7

### CODE REVIEW

File: builtin/logical/transit/path_encrypt.go [hunk 1]
```diff
@@ -126,7 +126,7 @@ to the min_encryption_version configured on the key.`,
 			"partial_failure_response_code": {
 				Type: framework.TypeInt,
 				Description: `
-Ordinarily, if a batch item fails to encrypt due to a bad input, but other batch items succeed, 
+Ordinarily, if a batch item fails to encrypt due to a bad input, but other batch items succeed,
```
Comment (cipherboy) [inline builtin/logical/transit/path_encrypt.go:129]: I'd revert this change, I think this will be reflowed by the `path-help` endpoint and the newline removed, so you'll need the space in here to avoid the comma from going into the next word.

File: sdk/helper/keysutil/policy.go [hunk 2]
```diff
@@ -241,6 +241,9 @@ type KeyEntry struct {
 	// This is deprecated (but still filled) in favor of the value above which
 	// is more precise
 	DeprecatedCreationTime int64 `json:"creation_time"`
+
+	// Key entry certificate chain. If set, leaf certificate key matches the keyEntry 'key'
```
Comment (cipherboy) [inline sdk/helper/keysutil/policy.go:245]: Let's add a comment that for each certificate, the first element is the leaf perhaps?

File: sdk/helper/keysutil/policy.go [hunk 3]
```diff
@@ -2392,3 +2395,191 @@ func wrapTargetPKCS8ForImport(wrappingKey *rsa.PublicKey, preppedTargetKey []byt
...
+	case KeyType_ECDSA_P521:
+		return elliptic.P521(), nil
+	default:
+		return nil, errutil.InternalError{Err: fmt.Sprintf("key type '%s' does not have a curve parameter", p.Type)}
+	}
+}
+
+func (p *Policy) ValidateAndPersistCertificateChain(ctx context.Context, storage logical.Storage, keyVersion int, certificateChain []*x509.Certificate) error {
+	if len(certificateChain) == 0 {
+		return errutil.UserError{Err: "expected at least a certificate in the certificate chain"}
```
Comment (cipherboy) [inline sdk/helper/keysutil/policy.go:2477]:
```suggestion
		return errutil.UserError{Err: "expected at least one certificate in the certificate chain"}
```

File: sdk/helper/keysutil/policy.go [hunk 4]
```diff
@@ -2392,3 +2395,191 @@ func wrapTargetPKCS8ForImport(wrappingKey *rsa.PublicKey, preppedTargetKey []byt
...
+	}
+
+	if certificateChain[0].BasicConstraintsValid && certificateChain[0].IsCA {
```
Comment (cipherboy) [inline sdk/helper/keysutil/policy.go:2480]: Hmm...

Part of me wants to walk this back: what if a key in Transit is for a CA certificate, not a leaf certificate?

In that case, I think the check would be:

```
if len(certChain) > 1 && certChain[0].BasicConstraintsValid && certChain[0].IsCA && (!certChain[-1].BCV || !certChain[-1].IsCA) {
  ...
}
```

i.e., our first item is a CA and our last item is not, implying the cert chain was presented in incorrect order... 

Thoughts?

File: sdk/helper/keysutil/policy.go [hunk 5]
```diff
@@ -2392,3 +2395,191 @@ func wrapTargetPKCS8ForImport(wrappingKey *rsa.PublicKey, preppedTargetKey []byt
...
+		return errutil.UserError{Err: "certificate in the first position is not a leaf certificate"}
+	}
+
+	for _, certificate := range certificateChain[1:] {
+		if certificate.BasicConstraintsValid && !certificate.IsCA {
+			return errutil.UserError{Err: "provided certificate chain contains more than one leaf certificate"}
+		}
+	}
+
+	valid, err := p.validateCertificateKeyMatch(keyVersion, certificateChain[0].PublicKeyAlgorithm, certificateChain[0].PublicKey)
```
Comment (cipherboy) [inline sdk/helper/keysutil/policy.go:2490]: Or, maybe just this check is sufficient, since it is unlikely that there'd be confusion with parent key material given they should all be distinct keys? Just a thought...

File: builtin/logical/transit/path_certificates.go [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+			},
+		},
+		HelpSynopsis:    pathCreateCSRHelpSynopsis,
+		HelpDescription: pathCreateCSRHelpDescription,
+	}
+}
+
+func (b *backend) pathImportCertChain() *framework.Path {
+	return &framework.Path{
+		Pattern: "keys/" + framework.GenericNameRegex("name") + "/set_certificate",
```
Comment (cipherboy) [inline builtin/logical/transit/path_certificates.go:52]:
```suggestion
		Pattern: "keys/" + framework.GenericNameRegex("name") + "/set-certificate",
```

Based on https://developer.hashicorp.com/vault/api-docs/secret/transit#set-certificate-chain :-)

File: builtin/logical/transit/path_certificates.go [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+				Type:        framework.TypeString,
+				Required:    true,
+				Description: "PEM encoded certificate chain. It should be composed by one or more concatenated PEM blocks and ordered starting from the end-entity certificate.",
+			},
+		},
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.UpdateOperation: &framework.PathOperation{
+				Callback: b.pathImportCertChainWrite,
+				DisplayAttrs: &framework.DisplayAttributes{
+					OperationVerb: "write",
```
Comment (cipherboy) [inline builtin/logical/transit/path_certificates.go:74]:
```suggestion
					OperationVerb: "set-chain",
```

perhaps?

File: builtin/logical/transit/path_certificates.go [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+				Type:        framework.TypeString,
+				Required:    false,
+				Description: "PEM encoded CSR template to use. The information attributes will be used as a basis for the CSR with the key in transit. If not set, a default template will be used.",
+			},
+		},
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.UpdateOperation: &framework.PathOperation{
+				Callback: b.pathCreateCSRWrite,
+				DisplayAttrs: &framework.DisplayAttributes{
+					OperationVerb: "write",
```
Comment (cipherboy) [inline builtin/logical/transit/path_certificates.go:41]:
```suggestion
					OperationVerb: "get-csr",
```

perhaps?

**Review by cipherboy — COMMENTED** [review]
Some initial comments, but I think this looks fairly good since this already landed in Vault (and thank you for porting your PR here from Vault!).

File: builtin/logical/transit/path_encrypt.go (hunk 1 above)
Comment (Gabrielopesantos) [reply builtin/logical/transit/path_encrypt.go]: Yes, sure. This was automatically done by the editor and I missed it when staging.

**Comment by cipherboy** [comment]
@Gabrielopesantos Also, you have two commits with bad DCOs, mind fixing that? Thanks!

https://github.com/openbao/openbao/pull/536/checks

> Summary
>
> Commit sha: [20f8fea](https://github.com/openbao/openbao/pull/536/commits/20f8fea1a114040afef07ef6c0bf4b3112228f59), Author: Gabriel Santos, Committer: Gabriel Santos; The sign-off is missing.
> Commit sha: [0e40ae9](https://github.com/openbao/openbao/pull/536/commits/0e40ae919f39a98995cf537151ffc34fb3909a0b), Author: Gabriel Santos, Committer: Gabriel Santos; The sign-off is missing.

File: builtin/logical/transit/path_certificates.go (hunk 7 above)
Comment (Gabrielopesantos) [reply builtin/logical/transit/path_certificates.go:74]: I see, I thought this wasn't arbitrary but re-read the field description. Your suggestion sounds good.

File: sdk/helper/keysutil/policy.go (hunk 4 above)
Comment (Gabrielopesantos) [reply sdk/helper/keysutil/policy.go:2480]: > Part of me wants to walk this back: what if a key in Transit is for a CA certificate, not a leaf certificate?

If this is the case, the call below would have to be somewhat different and check if there's a match between the key version in transit and a key in the chain. 
> 	valid, err := p.validateCertificateKeyMatch(keyVersion, certificateChain[0].PublicKeyAlgorithm, certificateChain[0].PublicKey)

With that, I'm not sure if I'm understanding your comment. Is it suggesting an additional check to make sure there's a CA certificate in the last position?

Comment (cipherboy) [reply sdk/helper/keysutil/policy.go:2480]: I was envisioning you'd always have the leaf-most CA's chain here, never a child CA/leaf cert if you had the CA's key. i.e., root -> ... -> leaf CA. ..

But thinking about this more, I'm not sure you'd have much value in a CA certificate that doesn't issue leaves (and thus, would be better suited for the PKI engine)... Unless you were issuing a type of certificate that Transit supported but PKI didn't and/or implementing CA software yourself. But then you'd probably have some other CA cert storage location. The one exception to this is a lot of self-signed service certs default to `IsCA=True` even if they're just TLS cets, so I could see that happening, but they should be able to change their self-signed creation step to set `IsCA=false` and then import it. 

So, let's keep this to leaves only is my thought!

File: sdk/helper/keysutil/policy.go [hunk 9]
```diff
@@ -2392,3 +2395,201 @@ func wrapTargetPKCS8ForImport(wrappingKey *rsa.PublicKey, preppedTargetKey []byt
...
+	if certificateChain[0].IsCA && certificateChain[0].BasicConstraintsValid {
+		return errutil.UserError{Err: "certificate in the first element is not a valid leaf certificate"}
+	}
+
+	// validate that the last element in the certificate chain is a CA certificate
+	if len(certificateChain) > 1 && !certificateChain[len(certificateChain)-1].IsCA {
+		return errutil.UserError{Err: "certificate in the last element is not a valid CA certificate"}
+	}
+
+	// NOTE: Does this check make sense?
```
Comment (cipherboy) [inline sdk/helper/keysutil/policy.go:2491]: Yes, the below check makes sense!

File: sdk/helper/keysutil/policy.go [hunk 10]
```diff
@@ -2392,3 +2395,201 @@ func wrapTargetPKCS8ForImport(wrappingKey *rsa.PublicKey, preppedTargetKey []byt
...
+		return elliptic.P384(), nil
+	case KeyType_ECDSA_P521:
+		return elliptic.P521(), nil
+	default:
+		return nil, errutil.InternalError{Err: fmt.Sprintf("key type '%s' does not have a curve parameter", p.Type)}
+	}
+}
+
+func (p *Policy) PersistCertificateChain(ctx context.Context, storage logical.Storage, keyVersion int, certificateChain []*x509.Certificate) error {
+	// validate that the certificate chain has at least one certifiate, the leaf certificate
```
Comment (cipherboy) [inline sdk/helper/keysutil/policy.go:2476]: Truncated comment here

File: sdk/helper/keysutil/policy.go [hunk 11]
```diff
@@ -2392,3 +2395,201 @@ func wrapTargetPKCS8ForImport(wrappingKey *rsa.PublicKey, preppedTargetKey []byt
...
+		return errutil.UserError{Err: "expected at least one certificate in the certificate chain"}
+	}
+
+	// validate that the first element in the certificate chain is a leaf certificate
+	if certificateChain[0].IsCA && certificateChain[0].BasicConstraintsValid {
+		return errutil.UserError{Err: "certificate in the first element is not a valid leaf certificate"}
+	}
+
+	// validate that the last element in the certificate chain is a CA certificate
+	if len(certificateChain) > 1 && !certificateChain[len(certificateChain)-1].IsCA {
```
Comment (cipherboy) [inline sdk/helper/keysutil/policy.go:2487]:
```suggestion
	if len(certificateChain) > 1 && (!certificateChain[len(certificateChain)-1].BasicConstraintsValid || certificateChain[len(certificateChain)-1].IsCA) {
```

File: sdk/helper/keysutil/policy.go [hunk 12]
```diff
@@ -2392,3 +2395,201 @@ func wrapTargetPKCS8ForImport(wrappingKey *rsa.PublicKey, preppedTargetKey []byt
...
+	case KeyType_RSA2048, KeyType_RSA3072, KeyType_RSA4096:
+		if certificatePublicKeyAlgorithm == x509.RSA {
+			keyTypeMatches = true
+		}
+	}
+	if !keyTypeMatches {
+		return false, errutil.UserError{Err: fmt.Sprintf("provided leaf certificate public key algorithm '%s' does not match the transit key type '%s'", certificatePublicKeyAlgorithm, p.Type)}
+	}
+
+	keyEntry, err := p.safeGetKeyEntry(keyVersion)
```
Comment (cipherboy) [inline sdk/helper/keysutil/policy.go:2556]: It seems weird to call this twice, tbh. I wonder if we could call it and pass it into this function from `PersistCertificateChain`?

**Comment by cipherboy** [comment]
@Gabrielopesantos A few very minor things and then I'll approve and merge! Thank you again!!

File: sdk/helper/keysutil/policy.go (hunk 10 above)
Comment (Gabrielopesantos) [reply sdk/helper/keysutil/policy.go:2476]: What do you mean? Should I split it into two lines?

File: sdk/helper/keysutil/policy.go (hunk 11 above)
Comment (Gabrielopesantos) [reply sdk/helper/keysutil/policy.go:2487]: Even if checking for `false`? I assumed that a check for `true` required checking the basic constraints, if not, it wasn't necessary.

File: sdk/helper/keysutil/policy.go (hunk 12 above)
Comment (Gabrielopesantos) [reply sdk/helper/keysutil/policy.go:2556]: Yes, that might be better.

File: sdk/helper/keysutil/policy.go (hunk 11 above)
Comment (Gabrielopesantos) [reply sdk/helper/keysutil/policy.go:2487]: In that case we might want the checks like the following, correct? If the BC value isn't valid, `true`, we can't trust the `IsCA` value so cannot properly evaluate the certificate "type".
```go
	// validate that the first element in the certificate chain is a leaf certificate
	if !certificateChain[0].BasicConstraintsValid || certificateChain[0].IsCA {
		return errutil.UserError{Err: "certificate in the first element is not a valid leaf certificate"}
	}

	// validate that the last element in the certificate chain is a CA certificate
	if len(certificateChain) > 1 && (!certificateChain[len(certificateChain)-1].BasicConstraintsValid || !certificateChain[len(certificateChain)-1].IsCA) {
		return errutil.UserError{Err: "certificate in the last element is not a valid CA certificate"}
	}

	// validate that the certificate chain contains only one leaf certificate
	for _, certificate := range certificateChain[1 : len(certificateChain)-1] {
		if !certificate.BasicConstraintsValid || !certificate.IsCA {
			return errutil.UserError{Err: "provided certificate chain contains more than one leaf certificate"}
		}
	}

```

File: sdk/helper/keysutil/policy.go (hunk 10 above)
Comment (cipherboy) [reply sdk/helper/keysutil/policy.go:2476]: Ohhh, I see. Gotcha :-)

File: sdk/helper/keysutil/policy.go (hunk 11 above)
Comment (cipherboy) [reply sdk/helper/keysutil/policy.go]: @Gabrielopesantos No, basic constraint not being valid implies that [it is _not_ a CA](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.9).

So the first check will also match on valid leaf certificates that don't have a basic constraint extension.

The second check is correct, because we want to fail all leaf certs: those without a basic constraint extension, or those with a basic constraint extension and IsCA asserted to true. The for loop's statement is correct for the same reason. :-)

Technically, the for loop duplicates the same as the second check, so we could probably just broaden the range to `certificateChain[1:len(certificateChain)]` perhaps?

File: sdk/helper/keysutil/policy.go [hunk 13]
```diff
@@ -2392,3 +2395,194 @@ func wrapTargetPKCS8ForImport(wrappingKey *rsa.PublicKey, preppedTargetKey []byt
...
+}
+
+func (p *Policy) PersistCertificateChain(ctx context.Context, storage logical.Storage, keyVersion int, certificateChain []*x509.Certificate) error {
+	// validate that the certificate chain has at least one certifiate, the leaf certificate
+	if len(certificateChain) == 0 {
+		return errutil.UserError{Err: "expected at least one certificate in the certificate chain"}
+	}
+
+	// validate that the first element in the certificate chain is a leaf certificate
+	if !certificateChain[0].BasicConstraintsValid || certificateChain[0].IsCA {
```
Comment (cipherboy) [inline sdk/helper/keysutil/policy.go:2482]:
```suggestion
	if certificateChain[0].BasicConstraintsValid && certificateChain[0].IsCA {
```

I think this is the only functional change left to make :-)

**Comment by cipherboy** [comment]
Interesting, I'd have expected different tests to fail. But the one failure looks like a server panic:

```

  2024-09-24T23:20:23.042Z [INFO]  TestTransit_Export_CertificateChain.core0.secrets.pki.pki_60706546: 03f9cb1e-8b26-cba1-96c8-995a6790476f: succeeded in migrating to issuer storage version 2
  2024-09-24T23:20:23.042Z [INFO]  TestTransit_Export_CertificateChain.core0.core: successful mount: namespace="" path=pki/ type=pki version=""
  2024-09-24T23:20:23.043Z [DEBUG] TestTransit_Export_CertificateChain.core0.secrets.pki.pki_60706546: Loading existing challenge validations on disk
  2024-09-24T23:20:23.043Z [DEBUG] TestTransit_Export_CertificateChain.core0.secrets.pki.pki_60706546: Starting ACME challenge validation engine
      path_export_test.go:681: 
          	Error Trace:	/home/runner/work/openbao/openbao/builtin/logical/transit/path_export_test.go:681
          	            				/home/runner/work/openbao/openbao/builtin/logical/transit/path_export_test.go:601
          	Error:      	Received unexpected error:
          	            	Put "https://127.0.0.1:37815/v1/transit/keys/rsa-2048/set-certificate": stream error: stream ID 15; INTERNAL_ERROR; received from peer
          	Test:       	TestTransit_Export_CertificateChain
  2024-09-24T23:20:45.236Z [INFO]  TestTransit_Export_CertificateChain: cleaning up vault cluster
  ```

File: sdk/helper/keysutil/policy.go [hunk 14]
```diff
@@ -2392,3 +2395,194 @@ func wrapTargetPKCS8ForImport(wrappingKey *rsa.PublicKey, preppedTargetKey []byt
...
+		return errutil.UserError{Err: "certificate in the first element is not a valid leaf certificate"}
+	}
+
+	// validate that the last element in the certificate chain is a CA certificate
```
Comment (cipherboy) [inline sdk/helper/keysutil/policy.go:2486]: Ah, I do think we should collapse these now.

```go
if len(certificateChain) > 1 {
	// validate that the certificate chain contains only one leaf certificate
	for _, certificate := range certificateChain[1 : len(certificateChain)] {
		if !certificate.BasicConstraintsValid || !certificate.IsCA {
			return errutil.UserError{Err: "could not validate if there is only one leaf certificate in the provided chain"}
		}
	}
}
```

This will fix the panic:

```
2024-09-25T11:26:06.715-0400 [INFO]  TestTransit_Export_CertificateChain: http2: panic serving 127.0.0.1:47784: runtime error: slice bounds out of range [1:0]
goroutine 663 [running]:
net/http.(*http2serverConn).runHandler.func1()
	/usr/local/go/src/net/http/h2_bundle.go:6185 +0x145
panic({0x2bfc8c0?, 0xc001624450?})
	/usr/local/go/src/runtime/panic.go:770 +0x132
github.com/openbao/openbao/sdk/v2/helper/keysutil.(*Policy).PersistCertificateChain(0xc001615900?, {0x362fbc0?, 0xc00161bec0?}, {0x3632620?, 0xc000bef630?}, 0x0?, {0xc000a922d0?, 0x0?, 0x0?})
	/home/ascheel/GitHub/cipherboy/openbao/sdk/helper/keysutil/policy.go:2492 +0x5ba
github.com/openbao/openbao/builtin/logical/transit.(*backend).pathImportCertChainWrite(0xc00032f030, {0x362fbc0, 0xc00161bec0}, 0xc0000fd860, 0xc001663190)
	/home/ascheel/GitHub/cipherboy/openbao/builtin/logical/transit/path_certificates.go:200 +0x5a6
github.com/openbao/openbao/sdk/v2/framework.(*Backend).HandleRequest(0xc000cec0e0, {0x362fbc0, 0xc00161bec0}, 0xc0000fd860)
	/home/ascheel/GitHub/cipherboy/openbao/sdk/framework/backend.go:293 +0xac8
github.com/openbao/openbao/vault.(*Router).routeCommon(0xc00088c140, {0x362fbc0, 0xc00161bec0}, 0xc0000fd860, 0x0)
	/home/ascheel/GitHub/cipherboy/openbao/vault/router.go:775 +0x153c
github.com/openbao/openbao/vault.(*Router).Route(...)
	/home/ascheel/GitHub/cipherboy/openbao/vault/router.go:552
github.com/openbao/openbao/vault.(*Core).doRouting(0xc00121b320?, {0x362fbc0?, 0xc00161bec0?}, 0xc000200690?)
	/home/ascheel/GitHub/cipherboy/openbao/vault/request_handling.go:823 +0x26
github.com/openbao/openbao/vault.(*Core).handleRequest(0xc00073a708, {0x362fbc0, 0xc00161bec0}, 0xc0000fd860)
	/home/ascheel/GitHub/cipherboy/openbao/vault/request_handling.go:964 +0xf73
github.com/openbao/openbao/vault.(*Core).handleCancelableRequest(0xc00073a708, {0x362fbc0, 0xc00161bec0}, 0xc0000fd860)
	/home/ascheel/GitHub/cipherboy/openbao/vault/request_handling.go:697 +0x11ff
github.com/openbao/openbao/vault.(*Core).switchedLockHandleRequest(0xc00073a708, {0x362fbc0, 0xc00161b620}, 0xc0000fd860, 0x1)
	/home/ascheel/GitHub/cipherboy/openbao/vault/request_handling.go:529 +0x605
github.com/openbao/openbao/vault.(*Core).HandleRequest(...)
	/home/ascheel/GitHub/cipherboy/openbao/vault/request_handling.go:482
github.com/openbao/openbao/http.request(0xc00073a708, {0x3620df0, 0xc00161b530}, 0xc0006bd0e0, 0xc0000fd860)
	/home/ascheel/GitHub/cipherboy/openbao/http/handler.go:819 +0x7f
github.com/openbao/openbao/http.handler.handleLogical.handleLogicalInternal.func57({0x3620df0, 0xc00161b530}, 0xc0006bd0e0)
	/home/ascheel/GitHub/cipherboy/openbao/http/logical.go:356 +0xa6
net/http.HandlerFunc.ServeHTTP(0xc00073a708?, {0x3620df0?, 0xc00161b530?}, 0xc00161bda0?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
github.com/openbao/openbao/http.handler.handleRequestForwarding.func35({0x3620df0, 0xc00161b530}, 0xc0006bd0e0)
	/home/ascheel/GitHub/cipherboy/openbao/http/handler.go:753 +0x143
net/http.HandlerFunc.ServeHTTP(0xc000634c30?, {0x3620df0?, 0xc00161b530?}, 0x6d7e88?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
net/http.(*ServeMux).ServeHTTP(0xc0008fe0c0?, {0x3620df0, 0xc00161b530}, 0xc0006bd0e0)
	/usr/local/go/src/net/http/server.go:2688 +0x1ad
github.com/openbao/openbao/http.handler.wrapHelpHandler.func47({0x3620df0, 0xc00161b530}, 0xc0006bd0e0)
	/home/ascheel/GitHub/cipherboy/openbao/http/help.go:28 +0xfd
net/http.HandlerFunc.ServeHTTP(0x0?, {0x3620df0?, 0xc00161b530?}, 0xc001115428?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
github.com/openbao/openbao/http.handler.wrapCORSHandler.func48({0x3620df0, 0xc00161b530}, 0xc0006bd0e0)
	/home/ascheel/GitHub/cipherboy/openbao/http/cors.go:32 +0x389
net/http.HandlerFunc.ServeHTTP(0xc00073a708?, {0x3620df0?, 0xc00161b530?}, 0xc000c3cd20?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
github.com/openbao/openbao/http.handler.rateLimitQuotaWrapping.func49({0x3620df0, 0xc00161b530}, 0xc0006bd0e0)
	/home/ascheel/GitHub/cipherboy/openbao/http/util.go:156 +0x9bf
net/http.HandlerFunc.ServeHTTP(0xc00161b440?, {0x3620df0?, 0xc00161b530?}, 0xc000ce2e70?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
github.com/openbao/openbao/http.wrapGenericHandler.func1({0x3620c70, 0xc000a92238}, 0xc000c15e60)
	/home/ascheel/GitHub/cipherboy/openbao/http/handler.go:421 +0xd35
net/http.HandlerFunc.ServeHTTP(0xc000c15d40?, {0x3620c70?, 0xc000a92238?}, 0xc00161b470?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
github.com/openbao/openbao/http.handler.wrapMaxRequestSizeHandler.func50({0x3620c70, 0xc000a92238}, 0xc000c15d40)
	/home/ascheel/GitHub/cipherboy/openbao/http/util.go:58 +0x122
net/http.HandlerFunc.ServeHTTP(0xc001658000?, {0x3620c70?, 0xc000a92238?}, 0x0?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
github.com/openbao/openbao/http.handler.PrintablePathCheckHandler.func51({0x3620c70, 0xc000a92238}, 0xc000c15d40)
	/home/ascheel/go/pkg/mod/github.com/hashicorp/go-cleanhttp@v0.5.2/handlers.go:42 +0x8f
net/http.HandlerFunc.ServeHTTP(0xd?, {0x3620c70?, 0xc000a92238?}, 0x524468?)
	/usr/local/go/src/net/http/server.go:2171 +0x29
net/http.serverHandler.ServeHTTP({0x79c900?}, {0x3620c70?, 0xc000a92238?}, 0xc000cec000?)
	/usr/local/go/src/net/http/server.go:3142 +0x8e
net/http.initALPNRequest.ServeHTTP({{0x362fbc0?, 0xc001456690?}, 0xc000f16388?, {0xc0005361e0?}}, {0x3620c70, 0xc000a92238}, 0xc000c15d40)
	/usr/local/go/src/net/http/server.go:3750 +0x231
net/http.(*http2serverConn).runHandler(0x4e47480?, 0x0?, 0x0?, 0xc0009b4000?)
	/usr/local/go/src/net/http/h2_bundle.go:6192 +0xbb
created by net/http.(*http2serverConn).scheduleHandler in goroutine 497
	/usr/local/go/src/net/http/h2_bundle.go:6127 +0x21d
2024-09-25T11:26:06.715-0400 [INFO]  TestTransit_Export_CertificateChain: cleaning up vault cluster
```

Comment (Gabrielopesantos) [reply sdk/helper/keysutil/policy.go:2486]: Nice catch. Forgot about this.

**Review by cipherboy — APPROVED** [review]
Thanks for your patience, @Gabrielopesantos! Merging :-)
